Jenkinsfile

```diff
@@ -0,0 +1,143 @@
+def isPR=false                   // true if the branch being tested belongs to a PR
+def baseProject=""               // name of base project - used when testing a PR
+def project=""                   // project where build and deploy will occur
+def projectCreated=false         // true if a project was created by this build and needs to be cleaned up
+def repoUrl=""                   // the URL of this project's repository
+
+// uniqueName returns a name with a 16-character random character suffix
+def uniqueName = { String prefix ->
+  sh "cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 16 | head -n 1 > suffix"
+  suffix = readFile("suffix").trim()
+  return prefix + suffix
+}
+
+// setBuildStatus sets a status item on a GitHub commit
+def setBuildStatus = { String url, String context, String message, String state, String backref ->
+  step([
+    $class: "GitHubCommitStatusSetter",
+    reposSource: [$class: "ManuallyEnteredRepositorySource", url: url ],
+    contextSource: [$class: "ManuallyEnteredCommitContextSource", context: context ],
+    errorHandlers: [[$class: "ChangingBuildStatusErrorHandler", result: "UNSTABLE"]],
+    statusBackrefSource: [ $class: "ManuallyEnteredBackrefSource", backref: backref ],
+    statusResultSource: [ $class: "ConditionalStatusResultSource", results: [
+        [$class: "AnyBuildResult", message: message, state: state]] ]
+  ]);
+}
+
+// getRepoURL retrieves the origin URL of the current source repository
+def getRepoURL = {
+  sh "git config --get remote.origin.url > originurl"
+  return readFile("originurl").trim()
+}
+
+// getRouteHostname retrieves the host name from the given route in an
+// OpenShift namespace
+def getRouteHostname = { String routeName, String projectName ->
+  sh "oc get route ${routeName} -n ${projectName} -o jsonpath='{ .spec.host }' > apphost"
+  return readFile("apphost").trim()
+}
+
+// Initialize variables in default node context
+node {
+  isPR        = env.BRANCH_NAME.startsWith("PR")
+  baseProject = env.PROJECT_NAME
+  project     = env.PROJECT_NAME
+}
+
+try { // Use a try block to perform cleanup in a finally block when the build fails
+
+  node ("maven") {
+
+    stage ('Checkout') {
+      checkout scm
+      repoUrl = getRepoURL()
+      stash includes: "ose3/pipeline-*.json", name: "artifact-template"
+    }
+
+    // When testing a PR, create a new project to perform the build 
+    // and deploy artifacts.
+    if (isPR) {
+      stage ('Create PR Project') {
+        setBuildStatus(repoUrl, "ci/app-preview", "Building application", "PENDING", "")
+        setBuildStatus(repoUrl, "ci/approve", "Aprove after testing", "PENDING", "") 
+        project = uniqueName("${appName}-")
+        sh "oc new-project ${project}"
+        projectCreated=true
+        sh "oc create serviceaccount jenkins -n ${project}"
+        sh "oc policy add-role-to-user edit -z jenkins -n ${project}"
+        sh "oc policy add-role-to-group edit system:authenticated -n ${project}"
```

Comment:
so the app artifacts use the 'jenkins' service account to run (so it has the permission to query other services, etc). Because with a PR project we don't have jenkins deployed, we need to create it explicitly and give it the right access. That's what the first 2 lines are for.

The third line is to allow access to the project to other users. Because it is the master project's jenkins service account that is creating it, no regular users have access to it. For this example, it is simply giving all authenticated users access, but it can be customized to a narrower set.